[feat] Add database constraints for validation annotations

Klaas058 opened the pull request:
Currently when using annotations with database rules like `Exists` and `Unique`, it is not possible to include database constraints.

This will throw a php error:
```php
// Error: Constant expression contains invalid operations
#[Exists(Customer::class, column: 'id', where: fn($query) => $query->where('active', true))]
public ?int $customer_id
 ```
 
 This feature would allow using database constraints like so:
 ```php
 // Valid php 🎉
#[Exists(Customer::class, column: 'id', where: new WhereConstraint('active', true))]
public ?int $customer_id
```

After testing this has been working great for us, happy with any feedback

One thing I'm not a 100% sure about:
* Types in the `DatabaseConstraint` implementations are `mixed` since the Laravel methods are untyped themselves. Is this the best way?

Klaas058 committed:
[feat] Add database constraints for validation annotations

rubenvanassche commented on docs/validation/using-validation-attributes.md:
````diff
@@ -222,6 +222,41 @@ The rules will now look like this:
+    ) {
+    }
+}
+```
````
Can you add a list here with the possible constraints?

rubenvanassche commented on src/Support/Validation/Constraints/WhereConstraint.php:
```diff
@@ -0,0 +1,17 @@
+{
+    public function __construct(
+        public readonly mixed $column,
+        public readonly mixed $value = null,
```
I know indeed laravel has a mixed type but they provide typehints, let's use those. For value mixed is allright since it probably is really mixed. Another thing I would like to see is support for ExternalReference since we can dynamically replace the value which might be useful.

Klaas058 replied:
Good suggestion! I've updated the PR to allow the use of `ExternalReference`. I also added types to the database constraints to the ones that I found made sense.

I took the type suggestions from https://api.laravel.com/docs/11.x/Illuminate/Validation/Rules/DatabaseRule.html 

I kept the `$value` of the `WhereConstraint` and `WhereNotConstraint` as `mixed`, like you said they probably really are mixed.

The type suggestion for `whereNot` does not include `null`, but that is supported.

This would otherwise error:
```php
new WhereNotConstraint('deleted_at', null)
```
Which is not a big deal, since you can use `WhereNullConstraint` instead, but it makes me think there are other inputs which are not in the type but are supported

rubenvanassche reviewed (changes requested):
Looking great, few changes required and then this one is good to go!

Klaas058 committed:
Merge branch 'main' of https://github.com/spatie/laravel-data into feature/annotation-validation-database-constraints

Klaas058 committed:
[feat] Allow use of ExternalReference in database constraints & add types to DatabaseConstraints

Klaas058 committed:
[feat] Add AppliesDatabaseConstraints trait & add list of database constraints to docs

Klaas058 committed:
[feat] Use mixed for value types for where and whereNot

rubenvanassche committed:
Merge branch 'main' into feature/annotation-validation-database-constraints

rubenvanassche committed:
Refactor database constraints to use self-applying pattern

- Replace Arrayable/toArray() with apply() method on DatabaseConstraint
- Simplify AppliesDatabaseConstraints trait to dispatch via polymorphism
- Add type docblocks for $where parameter on Exists and Unique
- Consistent constructor formatting across constraint classes
- Add missing standalone WhereNullConstraint test for Exists

rubenvanassche committed:
Merge branch 'main' into feature/annotation-validation-database-constraints

rubenvanassche committed:
Fix test failures caused by closure comparison in database rules

Laravel's whereIn/whereNotIn internally create closures stored in the
DatabaseRule $using array. These can't be compared via toEqual since
each call produces a new closure instance. Fixed by stringifying
database rules with query callbacks before comparison in assertRules.
Removed unfixable dataset entries from RulesDataset that hit the same
issue in the generic RulesTest which uses toEqual directly.

rubenvanassche committed:
Replace NormalizesExternalReferences trait with parseExternalReference method

Move the method to ValidationAttribute and DatabaseConstraint directly,
removing the trait indirection. Rename from
normalizePossibleExternalReferenceParameter to parseExternalReference.

rubenvanassche committed:
Keep normalizePossibleExternalReferenceParameter name in existing files

rubenvanassche commented:
Thanks!

MrMeshok commented:
```php
#[Exists(Customer::class, column: 'id', where: static function ($query) {$query->where('active', true);})]
public ?int $customer_id
```
This actually works in PHP 8.5 thanks to https://wiki.php.net/rfc/closures_in_const_expr. But note what you need to use `static function` not `fn`
